pkg/fqdn/rulegen.go

```diff
@@ -120,10 +126,26 @@ perRule:
 		// would lack the UUID-tagged rule and we would add a new UUID label in
 		// this function. Cleanup for existing rules with UUIDs is handled in
 		// StopManageDNSName
-		if !hasToFQDN(sourceRule) || sourceRule.Labels.Has(uuidLabelSearchKey) {
+		if !hasToFQDN(sourceRule) {
+			result = append(result, sourceRule)
 			continue perRule
 		}
 
+		// If the rule already have a UUID we check that it is present in the
+		// gen.allRules. If it is not present we don't need to append to
+		// result, due the rule was already updated by other process.
+		uuid := sourceRule.Labels.Get(uuidLabelSearchKey)
+		if uuid != "" {
+			_, exists := gen.allRules[uuid]
+			if !exists {
+				log.Debugf("Rule '%s' has been deleted when MarkToFQDNRules happens", uuid)
+			} else {
+				result = append(result, sourceRule)
+			}
+			continue perRule
```

Comment:
The commit does not describe why it is safe to omit the rest of the loop in the case when the UUID has not been found. It seems like a change in behavior that the commit message does not describe.